File: CpuA32/assembler_tool.py
#!/usr/bin/python3
"""
Assembler produces A32 ELF executables
"""
from typing import Dict, Any

import argparse
import os
import stat
import sys

from CpuA32 import assembler as a32


def lint(input):
    src = sys.stdin if input == "-" else open(input)
    print("UNIT", a32.UnitParse(src))


def assemble_common(input, output):
    src = sys.stdin if input == "-" else open(input)
    dst = sys.stdout if output == "-" else open(output, "wb")
    unit = a32.UnitParse(src)
    for sym in unit.symbols:
        assert sym.section, f"undefined symbol: {sym}"

    # print(unit)
    exe = a32.Assemble(unit, True)
    # for phdr in exe.segments:
    #    print(phdr)
    #    for sec in phdr.sections:
    #        print(sec)
    print("WRITING EXE")
    exe.save(dst)
    if output != "-":
        os.chmod(output, stat.S_IREAD | stat.S_IEXEC | stat.S_IWRITE)


def assemble(input, output):
    assemble_common(input, output)


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='assembler_tool')
    subparsers = parser.add_subparsers(dest='subparser')

    parser_lint = subparsers.add_parser('lint', description='just parse the input and print result')
    parser_lint.add_argument('input', type=str, help='input file')

    parser_assemble = subparsers.add_parser(
        'assemble',
        description='parse and emit elf exe. Also add _start entry point calling main')
    parser_assemble.add_argument('input', type=str, help='input file')
    parser_assemble.add_argument('output', type=str, help='output file')

    # First extract all the parser members into a dict
    kwargs: Dict[str, Any] = vars(parser.parse_args())
    # Next invoke the proper handler which is derived from the subparser
    # name. E.g. subparser `assembler_raw` is handled by the Python
    # function assembler_raw()
    handler = globals().get(kwargs.pop('subparser'))
    if handler:
        handler(**kwargs)
    else:
        parser.print_help(sys.stderr)

File: CpuA32/assembler_tool.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA32/assembler.h"

#include <fstream>
#include <string_view>

using namespace cwerg;

bool Assemble(std::string_view input, std::string_view output, bool add_startup_code) {
  std::ifstream finFile;
  std::istream* fin = &std::cin;
  if (input != "-") {
    finFile.open(input.data());
    fin = &finFile;
  }

  a32::A32Unit unit;
  if (!a32::UnitParse(fin, add_startup_code, &unit)) {
    std::cerr << "cannot parse input file " << input << "\n";
    return false;
  }
  auto exe = a32::MakeExe(&unit, true);
  std::cout << exe;

  std::ofstream foutFile;
  std::ostream* fout = &std::cout;
  if (output != "-") {
    foutFile.open(output.data());
    fout = &foutFile;
  }

  std::vector<std::string_view> chunks = exe.Save();

  for (const auto& c : chunks) {
    fout->write((const char*)c.data(), c.size());
  }
  return true;
}

int main(int argc, char* argv[]) {
  if (argc <= 1) {
    std::cerr << "no command specified\n";
    return 1;
  }

  if (argv[1] == std::string_view("lint")) {
    a32::A32Unit unit;
    a32::UnitParse(&std::cin, false, &unit);
    std::cout << unit << "\n";
  } else if (argv[1] == std::string_view("assemble_raw")) {
    if (argc <= 3) {
      std::cerr << "need src and dst args\n";
      return 1;
    }
    if (!Assemble(argv[2], argv[3], false)) {
      return 1;
    }
  } else if (argv[1] == std::string_view("assemble")) {
    if (argc <= 3) {
      std::cerr << "need src and dst args\n";
      return 1;
    }
    if (!Assemble(argv[2], argv[3], true)) {
      return 1;
    }
  }
  return 0;
}
